client/concordclient/include/client/concordclient/concord_client.hpp

```diff
@@ -130,14 +132,59 @@ class ConcordClient {
 
   // Register a callback that gets invoked for every validated event received.
   // void callback(EventGroup);
+  // Return subscriber ID used to unsubscribe.
   template <class CallbackT>
   void subscribe(const SubscribeRequest& request,
                  const std::unique_ptr<opentracing::Span>& parent_span,
                  CallbackT callback);
+
+  // Unsubscribe with the subscriber ID retrieved from a prior call to subscribe.
+  // Note, if the caller doesn't unsubscribe and no runtime error occurs then resources
+  // will be occupied forever.
   void unsubscribe();
 
  private:
+  logging::Logger logger_;
   ConcordClientConfig config_;
+
+  // TODO: Allow multiple subscriptions
+  std::atomic_bool stop_subscriber_{true};
+  std::thread subscriber_;
 };
 
+template <class CallbackT>
+void ConcordClient::subscribe(const SubscribeRequest& request,
+                              const std::unique_ptr<opentracing::Span>& parent_span,
+                              CallbackT callback) {
+  if (not stop_subscriber_) {
+    LOG_ERROR(logger_, "subscription already in progress - unsubscribe first");
+    return;
+  }
+
+  stop_subscriber_ = false;
+  std::atomic_bool* stop = &stop_subscriber_;
+  subscriber_ = std::thread([stop, request, callback] {
+    while (not(*stop)) {
+      // Note: The following returns an artificial event group.
+      // This will be replaced with the actual thin replica client integration.
+      // The thread is in place to simulate the asynchronous subscription.
+      EventGroup eg;
+      eg.id = request.event_group_id;
+      std::string event = std::to_string(request.event_group_id);
+      eg.events.push_back({event.begin(), event.end()});
+      std::chrono::duration time_now = std::chrono::system_clock::now().time_since_epoch();
+      eg.record_time = std::chrono::duration_cast<std::chrono::microseconds>(time_now);
+      eg.trace_context = {};
+
+      callback(eg);
+    }
+  });
+}
+
+inline void ConcordClient::unsubscribe() {
```

Comment:
Please remove `inline`:
1. `inline` only recommends to a compiler to make the function inlined, i.e. it is not guaranteed.
2. functions defined in headers are implicitly treated by the C++ compiler as candidates for inlining
2. a "heavy" function has no chance to become inline. This function is super-heavy because it calls `join`(interaction with the kernel).

Reply:
`inline` isn't only to tell the compiler about possible optimizations (which we don't need to do anyways) but also to work around the "one-definition-rule". Meaning, if we define a function in a header file (which we have to do because we are using templates) then multiple includes of this file in different compilation units can lead to double-definition when the linker links those objects together. This is exactly what happened here. We will need to make the other functions `inline` as well.
This ties back to your API remark above, if we would use `std::function` instead of templates then we wouldn't face this issue. However, so far, I've valued the benefits of templates higher then the issues I faced.
@andrewjstone might have insightful opinions too.

Reply:
Well, `inline` is still redundant: you could have written the method definition inside the class. In this case, the compiler will handle "one-definition-rule" for you implicitly. So now, I suggest putting this method to the class declaration.

Reply:
I don't like that because then you lose the overview of the class declaration. At a first glance, I don't care about the implementation but the API and the documentation. Therefore, I prefer for it to be as concise as possible.